### PR TITLE
EIP-crosschain-id proposal

### PR DESCRIPTION
**Pull request by weijia31415**
This document was created by EEA (Enterprise Ethereum Alliance) Crosschain Interoperablitty Task Force and is submitted as a EIP proposal.  The document was shared with Hudson Jameson for EIP editor's feedback.

When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:14]: The simple summary should really only be a couple sentences. The requirements specified here would make more sense in the motivation / rationale.

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:5]: Please update with URL to discuss this EIP. `https://ethereum-magicians.org` is preferred.

Comment (lightclient) [reply EIPS/eip-crosschain-id.md:5]: This EIP cannot be merged until this is fulfilled.

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:16]: Is this truly case? I believe `v` can be up to `2^256`. Since the chain id is multiplied by 2, that should leave `2^255` identifiers?

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:8]: If you're proposing a standard for signing messages that is not compatible with EIP-155, the category should be `Core`.

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:91]:
```suggestion
```

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:93]: Looks like a few awkward new lines here.

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:89]: This section is generally reserved to discuss why *specific* decisions were made regarding the specification. The motivation section is a better place for longer prose explanations of why the EIP is being specified.

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:99]: I don't think this is true. Ethereum clients today validate transactions by extracting the chain id from `v`. Even if there is a 4 byte segment to record the chainid, clients would need to know how to extract it. This would be a core change and necessitate a hard fork.

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:118]: Can probably drop this sentence as it is implied by the section title.

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:106]: This section is generally being deprecated, so if you don't have an implementation to provide here inline I recommend removing it altogether.

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:101]: Please provide concrete test cases for the EIP - otherwise it is recommended to remove this section.

File: EIPS/eip-crosschain-id.md
Comment (lightclient) [inline EIPS/eip-crosschain-id.md:116]: There are few formatting ideas in this section, it would be best to choose one format and stick with it. For example: 

```
problem: solution
or
#### problem
solution
```

**Review by lightclient — CHANGES_REQUESTED** [review]
I left a few recommendations on your PR. The main blocker is the discussion URL. I'd also like to better understand how you expect this to interoperate with mainnet. It seems like it would conflict with EIP-155, in which case you would need to submit a core EIP to support the new chainId format. If that is the case, you will need to create a separate ERC EIP defining the smart contract interface.

File: EIPS/eip-crosschain-id.md
Comment (MicahZoltu) [reply EIPS/eip-crosschain-id.md:116]: I recommend `###` rather than `####` to keep ToC tree neat.  😄

File: EIPS/eip-crosschain-id.md
Comment (MicahZoltu) [inline EIPS/eip-crosschain-id.md:131]:
```suggestion
```
I don't know anything about EEA copyrights, but I believe every EIP currently has the exact same copyright section and I'm a bit loath to have one be a bit different (for any reason).  If you are interested in discussing adding `EEA copyright waived` more generally to EIPs, that is a welcome discussion and I recommend opening an issue or a PR against the template.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md:12]: Sure. We will review in today's meeting and make changes.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md:14]: Yes, chainid is currently a 32byte (256 bit) field with EIP-1344.   The spec of EIP-155 gave a perception that chain-id is an integer.  This spec is to give semantic meaning to this 256 bit/32 byte chainid. 
We can modify this as well.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md:86]: Sure  Will make needed changes.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md:97]: We are getting test cases recommendation from EEA testnet team.  So will be good to keep it and add test cases.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md:102]: Sure.  Will remove this one.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md:114]: Will do that.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md:127]: This is from EEA legal team.  I will pass your comments to the legal team.

File: EIPS/eip-crosschain-id.md
Comment (MicahZoltu) [reply EIPS/eip-crosschain-id.md:2]: Filename needs to change to `eip-3220.md` as well.

Comment (weijia31415) [reply EIPS/eip-crosschain-id.md]: EIP number has been added and file name changed.  Thank you.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md]: We discussed this further in EEA crosschain taskforce meeting.  Crosschain id can be processed at the dapp level, smart contract level. If this is built into transaction level, then the core module will need to extract the legacy chainid from crosschain id.   Peter Robinson might comment more on this.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md]: Yes, changed email to github handles.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md]: Yes, dropped from the document.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md]: Discussed this in EEA Crosschain taskforce meeting.  Maybe leave this section for now as a place holder for discussions on implementations at the smart contract layer or core layer.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md]: Seems the format has been changed.

File: EIPS/eip-crosschain-id.md
Comment (weijia31415) [reply EIPS/eip-crosschain-id.md]: Yes, changed the document.

File: EIPS/eip-3220.md
Comment (lightclient) [inline EIPS/eip-3220.md:15]: The simple summary should generally just be 1-2 sentences. I recommend moving this to motivation.

File: EIPS/eip-3220.md
Comment (lightclient) [inline EIPS/eip-3220.md:29]:
```suggestion
```

**Review by lightclient — CHANGES_REQUESTED** [review]
Hello, this draft is mainly blocked on a `discussions-to` field in the header. I recommend opening a thread on https://ethereum-magicians.org/ for this EIP.

File: EIPS/eip-3220.md
Comment (weijia31415) [reply EIPS/eip-3220.md]: Modified and move this content to Motivation

File: EIPS/eip-3220.md
Comment (MicahZoltu) [inline EIPS/eip-3220.md:14]: Simple Summary should be shorter, think subject line of an email or title line of a GitHub pull request or forum thread.
```suggestion
A self-verifying unique blockchain identifier that deals with forks.
```

File: EIPS/eip-3220.md
Comment (MicahZoltu) [inline EIPS/eip-3220.md:18]: The first half of this is more appropriate for motivation section.  The abstract should be a human readable high level overview of the specification.  It shouldn't go into the "why" at all (that is for the motivation section).
```suggestion
The crosschain-id is a 32 byte hex string and with some bytes extracted from blockchain hash and some manually defined to characterize a blockchain.
We also propose a registration and lookup service to retrieve blockchain metadata from the crosschain-id.
```
You may want to go into a *little* more detail, maybe one or two sentences or one line that just shows the bit layout.

File: EIPS/eip-3220.md
Comment (MicahZoltu) [inline EIPS/eip-3220.md:22]: Recommend leaving proposer name out and just reference EIP-155 (no need to appeal to authority).

Comment (MicahZoltu) [reply EIPS/eip-3220.md:22]:
```suggestion
[EIP-155](./eip-155.md) provides a unique identifier to a blockchain to provide simple relay attack protection.  This specification defines an integer for Chainid for a blockchain and sign the chainid into a transaction data and hence present attackers to send same transaction to different blockchains. This specification will require blockchains to define a chainid and register the chainid in a public repository.
```

Comment (MicahZoltu) [reply EIPS/eip-3220.md:22]: > and sign the chainid into a transaction data and hence present attackers to send same transaction to different blockchains

This sentence doesn't seem to be properly structured.  I'm unable to figure out what you are trying to say.

File: EIPS/eip-3220.md
Comment (MicahZoltu) [inline EIPS/eip-3220.md:36]: Recommend a bulleted list.  In markdown this will render as a paragraph, without proper capitalization.

File: EIPS/eip-3220.md
Comment (MicahZoltu) [inline EIPS/eip-3220.md:41]:
```suggestion
```
I think we can remove this as it doesn't really add much value.  The style of an EIP specification should not be like an essay where you have an introduction, body, conclusion.  Just focus on the technical specification only.

File: EIPS/eip-3220.md
Comment (MicahZoltu) [inline EIPS/eip-3220.md:48]: Recommend an anchor link to the table being referred to.

File: EIPS/eip-3220.md
Comment (MicahZoltu) [inline EIPS/eip-3220.md:81]:
```suggestion
```
Recommend making this a separate EIP. It doesn't need to be part of this EIP and instead can just reference this EIP as a dependency. Also, this would be an ERC while the chain ID definition would (arguably) be a Core EIP (if signing uses it) or an interface EIP (if it is just a conceptual standard).

File: EIPS/eip-3220.md
Comment (MicahZoltu) [inline EIPS/eip-3220.md:87]: Very weird newlines here.  Recommend a newline after every sentence.  Markdown will turn that into one paragraph, and it makes it much easier to give targeted feedback on specific lines.

File: EIPS/eip-3220.md
Comment (MicahZoltu) [inline EIPS/eip-3220.md:88]: This should all be moved to the motivation section.  The Rationale section is for explaining why you made specific technical choices.  For example, why is the truncated block hash 16 bytes instead of 17 bytes, why is chainId 4 bytes instead of 8 bytes, etc.

File: EIPS/eip-3220.md
Comment (MicahZoltu) [inline EIPS/eip-3220.md:108]:
```suggestion
```
These sections are optional and should be removed unless there is inline or attached tests or a reference implementation.

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
While none of the changes proposed here are *required* for a merge to draft, many of them will block a merge to Review status.  I'm going to wait a day to see if the authors would like to make those changes now (before merging as Draft).

File: EIPS/eip-3220.md
Comment (weijia31415) [reply EIPS/eip-3220.md]: Modified

File: EIPS/eip-3220.md
Comment (weijia31415) [reply EIPS/eip-3220.md]: Modified

File: EIPS/eip-3220.md
Comment (weijia31415) [reply EIPS/eip-3220.md]: Modified and moved the descriptive ones to Rational.

**Pull request merged**
